File: Rotate Array.py
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        
        k %= len(nums)
        
        if k:
            start = 0
            end = len(nums)
            reverse = False
            while(end - start > k):
                if 2 * k > (end - start):
                    k = end - start - k
                    reverse = not(reverse)
                if not(reverse):
                    for i in range(k):
                        nums[start + i], nums[end - k + i] = nums[end - k + i], nums[start + i]
                    start += k
                else:
                    for i in range(1, k+1):
                        nums[end-i], nums[start+k-i] = nums[start+k-i], nums[end-i]
                    end -= k

File: Rotate Array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int maxIdx = nums.size();
        k %= maxIdx;
        if (k == 0) return;
        
        int start = 0;
        int curIdx = start;
        int nextIdx = start + k;
        int curTemp = nums[curIdx];
        int nextTemp;
        
        for (int dummy = 0; dummy < maxIdx; ++dummy)
        {
            nextTemp = nums[nextIdx];
            nums[nextIdx] = curTemp;
            curTemp = nextTemp;
            curIdx = (curIdx + k) % maxIdx;
            nextIdx = (nextIdx + k) % maxIdx;
            if (curIdx == start)
            {
                start += 1;
                curIdx = start;
                curTemp = nums[curIdx];
                nextIdx = start + k;
            }
        }
    }
};
